### PR TITLE
introduce config.h for compile definitions

### PR DESCRIPTION
**Pull request by tjhei**
this removes the need for COMPILE_DEFINITIONS, which interferes with
UNITY_BUILD.

see #3585

### CODE REVIEW

**Comment by tjhei** [comment]
One thing I am a bit worried about is that we are not including global.h in every header/source file at the moment. Is this something we should enforce and maybe check with the "indent" check? Otherwise the defines will not be available in the files that forget to include it...

File: cmake/AspectConfig.cmake.in [hunk 1]
```diff
@@ -41,27 +41,6 @@ MACRO(ASPECT_SETUP_PLUGIN _target)
     INCLUDE_DIRECTORIES "${Aspect_INCLUDE_DIRS}"
   )
 
-  IF(ASPECT_USE_PETSC)
-    MESSAGE(STATUS "  with PETSC=ON")
-    SET_PROPERTY(TARGET ${_target}
-      APPEND PROPERTY COMPILE_DEFINITIONS ASPECT_USE_PETSC="1")
-  ENDIF() 
-
-  IF(ASPECT_USE_PETSC)
```
Comment (gassmoeller) [inline cmake/AspectConfig.cmake.in:50]: This was wrong all the time? :smile:

Comment (tjhei) [reply cmake/AspectConfig.cmake.in:50]: yes. :-)

**Comment by tjhei** [comment]
> Yes I think its reasonable to expect every file to include "global.h". Its implied in the name.

I am not sure if it should be a) every .h, b) every .cc, or c) both.

> Do you want to make the indent check part of this PR?

How about we do that separately? This PR seems to work as is.

**Comment by gassmoeller** [comment]
> I am not sure if it should be a) every .h, b) every .cc, or c) both.

Hm, thats complicated you are right. It would also mean a lot of duplicated includes. I just checked: `global.h` is included from `plugins.h` (included in all the interfaces), `simulator_access.h`, `simulator.h`, `simulator_signals.h` and `utilities.h`. So maybe we do not need to worry about the check at all. I can not imagine how one would write a source file for ASPECT without including any of these headers.

> How about we do that separately? This PR seems to work as is.

You are right. I will merge this and we can think about the check separately.

**Comment by tjhei** [comment]
> I can not imagine how one would write a source file for ASPECT without including any of these headers.

Well, you can use a macro before including one of the ASPECT headers (see the world builder changes in this PR).

**Comment by gassmoeller** [comment]
True, but we can not really come up with a test for that. We would need to check if there is any use of a macro before an include of global.h, which seems impractical. Also this should be noticed during regular testing.

**Pull request merged**